documentation/client-side-forms/appendix-client-side-form-binding-reference.asciidoc

```diff
@@ -0,0 +1,348 @@
+
+There are the builtin models that represent the common primitive field types:
+
+|===
+| Type | Value type `T` | Empty value (`Model.createEmptyValue()` result)
+
+| `StringModel` | `string` | `''`
+| `NumberModel` | `number` | `0`
+| `BooleanModel`  | `boolean` | `false`
+|===
+
+Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.
+
+=== Object models
+
+Typically, the primitive editable values of the form are grouped in objects. `ObjectModel<T>` accommodates that. `ObjectModel<T>` is also a common superclass for all the models generated from Java beans.
+
+The subclasses of `ObjectModel<T>` define the type argument constrains and the default type. In addition, the subclasses list all the public properties, following the shape of the described object.
+
+For example, for the following Java bean:
+
+.IdEntity.java
+[source, java]
+----
+public class IdEntity {
+    private String idString;
+
+    public String getIdString() {
+        return idString;
+    }
+
+    public void setIdString(String idString) {
+        this.idString = idString;
+    }
+}
+----
+
+.Person.java
+[source,java]
+----
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.NotEmpty;
+
+public class Person extends IdString {
+    @NotEmpty(message = "Cannot be empty")
+    private String fullName;
+
+    public String getFullName() {
+        return fullName;
+    }
+
+    public void setFullName(String fullName) {
+        this.fullName = fullName;
+    }
+}
+----
+
+The following TypeScript interfaces are generated for type-checking endpoints:
+
+.IdEntity.ts
+[souces, typescript]
+----
+export default interface IdEntity {
+  idString: string;
+}
+----
+
+.Person.ts
+[source, typescript]
+----
+import IdEntity from './IdEnity';
+
+export default interface Person extends IdEntity {
+  fullName: string;
+}
+----
+
+And the following models are generated for client-side form binding:
+
+.IdEntityModel.ts
+[source, typescript]
+----
+import IdEntity from './IdEntity';
+
+export default class IdEntityModel<T extends IdEntity = IdEntity> extends ObjectModel<T> {
+  static createEmptyValue: () => IdEntity;
+  readonly idString = new StringModel(this, 'idString');
+}
+----
+
+.PersonModel.ts
+[source, typescript]
+----
+import IdEntityModel from './IdEntityModel';
+
+import Person from './Person';
+
+export default class PersonModel<T extends Person = Person> extends IdEntityModel<T> {
+  static createEmptyValue: () => Person;
+  readonly fullName = new StringModel(this, 'fullName', new NotEmpty({message: 'Cannot be empty'}));
+}
+----
+
+[IMPORTANT]
+====
+To avoid naming collisions with user-defined object model fields, the builtin models and model superclasses do not have any public instance properties or methods, aside form the `toString` and `valueOf` exceptions mentioned above.
+====
+
+[NOTE]
+====
+The properties of object models are intentionally read-only.
+====
```

Comment:
Not sure if all these admonition blocks are necessary (across the entire article, not these two in particular)? It might make it harder to read the article if they are used too frequently.

Reply:
@magi42, do you have any guidelines when admonitions should be used?

Reply:
I’ve inlined some of them here.

Reply:
Okay, great! Good if you could identify some that don’t benefit from being called out more. I don’t really have any concrete guidance for these, just that too much of those will make it hard to read :)